module4-authentication-and-security/securing-backend-app/README.md

```diff
@@ -0,0 +1,79 @@
+
+## Let's add some security
+
+To secure our wishlist above, we can add a new column to the lists table to attach it to an identity (a user). We can name that column `user_email` for example, and use this query to get the user wishlist:
+
+```SQL
+SELECT `id`, `name`, `user_email` from `lists`
+INNER JOIN `items`
+ON `items`.`list_id` = `lists`.`id`
+WHERE `lists`.`user_email` = "test@domain.com"
+```
+
+Thus, when a user comes to your app, you can ask them first about their email address, and then query the database to get thier lists. But that is also insufficient. A malicious user can peep on other users wishlists if he/she knew thier emails, or he/she check different email combinations to see if it lands a hit, aka, **Brute Force**.
+
+To make this a bit harder to brute force, you can create a token system. Tokens are random strings of charcters, usually wih arbitrary length, that is intended to be hard to brute force (guess). A token can look something like this: `1FsahNjuio31OHsOallsn1gHH7`
+
+You can change the column `user_email` to `token` and everytime a user comes in, you can ask them for the token, or give them the opportunity to generate a token. If they are new users, they can generate a new token, and save it, then start creating thier lists. When they visit again, they enter that token to get thier lists back.
+
+Bingo! You've just created your first, very acceptable authentication layer! You've secured your app and **slimmed the chances** of brute forcing user data! Many apps today use this pattern to secure data, however it is not always acceptable. That's because, if a user lost thier token, it is next to impossible (especially if the token was generated by a cryptograpic random generators with reasonable length) to to gain their access back to their lists using today's technologies.
+
+## Why we need passwords
+
+The token system developed above is very robust, but it isn't always acceptable security measure (at least not alone). That's because, if a user lost thier token, it is next to impossible (especially if the token was generated by a cryptograpic random generators with reasonable length) to to gain their access back to their lists using today's technologies.
+
+So, maybe instead of a long random token, use a password? We can ask the user when they land on our page for a pass code (aka password). And we check the database, if we can find it a list where `password = USER_SUBMITTED_PASS_CODE` then display them. If not, that means they are a new user.
+
+Indeed this works, but we are back to the same problem with using emails. Users tend to use easy to remember words, dates, phone numbers, and the like. It would be relatively harder than bruteforcing emails, but nonetheless, still easy to guess, and very easy for computers to break by checking every possible combination of words, dates, numbers, and even relatively harder passwords using a technology called **rainbow tables**.
```

Comment:
there is no explanation of rainbow tables? is that coming later? i think you should definitely put TODOs if you are planning to expand certain topics

Reply:
also i don't think this is the same problem as using emails, so the first sentence on line 62 is also misleading.

there's nothing wrong with passwords in general. i would suggest writing this section as follows

passwords are okay and a perfectly viable and widespread approach, but there are two things to remember about passwords
1. require validation to prevent brute force attacks (thus preventing easy to guess passwords)
2. use salt and hash to prevent rainbow attacks -- point to a code example here or include your own, and then this needs an in-depth explanation because this is probably the most important section that new people will forget about writing passwords

these are the two essential things about password security, you can get away without knowing much else (and i think there's not much else to know here)